File: Lecture24/infwell.cpp
#include "cptstd.hpp"
#include "basalg.hpp"
#include "diffeq.hpp"
using namespace cpt;

class ParticleInBox {

public : 
  enum extended_space { _x, _psi, _dpsi_dx, dimension };


  ParticleInBox( double acc ) : accuracy(acc) { E = 0.0;}

protected : 
  double E;                       // current value of E for root finding
  double accuracy;                // for integration and root finding
};


class ParticleInBoxEnergy : public ParticleInBox {
public : 

  ParticleInBoxEnergy( double iE, double acc ) : ParticleInBox( acc), E(iE) {}

  double V(double x) const {            // potential function
    if (x >= 0 && x <= 1)
      return 0;
    else return 1e30;           // large value
  }

  // flow vector for Runge-Kutta integration
  Matrix<double,1> operator()(Matrix<double,1>& p)  const       // point in extended space
  {
    double x = p[_x], psi = p[_psi], dpsi_dx = p[_dpsi_dx];
    Matrix<double,1> flow(dimension);
    flow[_x] = 1;
    flow[_psi] = dpsi_dx;
    flow[_dpsi_dx] = (V(x) - E) * psi;
    return flow;
  }
protected : 
  double E;

};

class ParticleInBoxWavefunction : public ParticleInBox {
public : 
  ParticleInBoxWavefunction( double acc ) : ParticleInBox( acc) {}

  double operator()(double iE)   const           // search function for root finding
  {

    ParticleInBoxEnergy energy( iE, accuracy );

    Matrix<double,1> p(dimension);
    p[_x] = 0;                  // left wall
    p[_psi] = 0;                // psi(0)
    p[_dpsi_dx] = 1;            // arbitrary value of slope
    double dx = 0.001;          // suggested step size for Runge-Kutta
    double Delta_x = 1.0;       // integration interval
    RK4_integrate(p, dx, energy, Delta_x, accuracy);

    return p[_psi];             // looking for psi(1) = 0
  }

  double solve(double E0, double E1) {
    // use bisection search to find the energy eigenvalue
    BisectionSearchT<ParticleInBoxWavefunction> bs;                 // Bisection search object
    bs.set_first_root_estimate(E0);
    bs.set_second_root_estimate(E1);
    bs.set_accuracy(accuracy);
    bs.print_steps();                   // root finder will print each step
    bs.find_root(*this);
  }

};

int main() {

    cout << " Bound state energies of infinitely deep potential well\n"
         << " ------------------------------------------------------\n"
         << " Enter bracketing guesses E_0, E_1, and desired accuracy: ";
    double E0, E1;
    double accuracy;
    cin >> E0 >> E1 >> accuracy;

    ParticleInBoxWavefunction infwell(accuracy);
    double E = infwell.solve(E0, E1);
    cout << " Eigenvalue E = " << E << endl;

    // integrate to get wavefunction
    ofstream data_file("infwell.data");
    int N = 100;
    double dx = 1.0 / N;
    Matrix<double,1> p(ParticleInBox::dimension);
    p[ParticleInBox::_x] = 0, p[ParticleInBox::_psi] = 0, p[ParticleInBox::_dpsi_dx] = 1;
    ParticleInBoxEnergy energy(E,accuracy);
    data_file << p[ParticleInBox::_x] << '\t' << p[ParticleInBox::_psi] << '\t' << p[ParticleInBox::_dpsi_dx] << '\n';
    for (int i = 0; i < N; i++) {
        RK4_step(p, dx, energy);          // take one Runge-Kutta step
        data_file << p[ParticleInBox::_x] << '\t' << p[ParticleInBox::_psi] << '\t' << p[ParticleInBox::_dpsi_dx] << '\n';
    }
    data_file.close();
    cout << "\n " << N << " step solution in file infwell.data"
         << endl;
}

File: Lecture24/infwell.py
import math
import cpt


class ParticleInBox :
    E = 0.0             # current value of E for root finding
    accuracy = 1.0e-6   # accuracy for adaptive RK4 integration
    def __init__( self, E, accuracy ) :
        self.E = E
        self.accuracy = accuracy


class ParticleInBoxEnergy( ParticleInBox ) :
    def __init__( self, E, accuracy ):
        ParticleInBox.__init__( self, E, accuracy )
    def __call__(self, xy):   # time independent Schroedinger equation
        x = xy[0] ; psi = xy[1] ; psi_prime = xy[2]
        dx_dx = 1.0
        dpsi_dx = psi_prime
        dpsi_prime_dx = ( self.V(x) - self.E ) * psi
        return [ dx_dx, dpsi_dx, dpsi_prime_dx ]
    def V(self, x):           # potential function
        if x >= 0.0 and x <= 1.0:
            return 0.0
        else:
            return 1.0e30

class ParticleInBoxWavefunction( ParticleInBox ) :
    def __init__( self, E, accuracy ):
        ParticleInBox.__init__( self, E, accuracy )
        self.particleE = ParticleInBoxEnergy( E, accuracy)
    def __call__(self, energy):      # search function for root finding
        self.E = energy
        self.particleE.E = energy
        x = 0.0         # left wall
        psi = 0.0       # psi(0)
        psi_prime = 1.0 # arbitrary value of slope
        xy = [ x, psi, psi_prime ]
        Delta_x = 1.0   # width of well
        dx = 0.001      # suggested step size of Runge-Kutta
        cpt.RK4_integrate(xy, dx, self.particleE, Delta_x, self.accuracy)
        return xy[1]    # searching for psi(x=1) = 0



print " Bound state energies of infinitely deep potential well"
print " ------------------------------------------------------"
E0, E1, accuracy = input(" Enter guess E_0, E_1, and desired accuracy: ")
particlePsi = ParticleInBoxWavefunction( E0, accuracy )
E = cpt.root_bisection(particlePsi, E0, E1, particlePsi.accuracy, 1000, True)
print " Eigenvalue =", E
